README.md
# DispatchMessage2Listeners

* Author: Renato Miawaki

When you need to dispatch events to methods (listeners) and filter if message is equal.

For socket stream you can use to separate messages and distribute just for clients who want to get the message

## For use in stream

```

const DispatchMessage2Listeners = require('./index') ;

const NmeaBuffer = require( 'nmea-buffer' ) ;

//Example 1, parsing NMea messages
const n1 = new DispatchMessage2Listeners( 10, NmeaBuffer.parse ) ;
n1.showLog = true ;
n1.addListener("a", (d)=>{
    console.log("recived a in n1", d.message) ;
} ) ;

n1.addListener("B", (d)=>{
    console.log("recived B in n1", d.message) ;
} ) ;
n1.addListenerAll((d)=>{
    console.log("recived ALL in n1", d.message) ;
} ) ;
n1.addListenerAllOnChange((d)=>{
    console.log("recived ALL on change in n1", d.message) ;
} ) ;
let data = [ ... NmeaBuffer.getMessageBuffer("a", "just for a 1"), ...NmeaBuffer.getMessageBuffer("a", "just for a 2"), ...NmeaBuffer.getMessageBuffer("B", "just for Bs"), ...NmeaBuffer.getMessageBuffer("c", "just for c") ];

n1.parseData(data) ;


//Exemple 2 - parsing using json with method and body attributes
const n2 = new DispatchMessage2Listeners( 10, (data)=>{
    let parsedString = JSON.parse( data.toString() ) ;
    return { header:parsedString.method, message:parsedString.body }
} ) ;
n2.addListener("a",  (d)=>{
    console.log("recived a in n2", d) ;
} )

n2.addListenerOnChange("B",  (d)=>{
    console.log("recived B in n2", d) ;
} )
const Buffer        = require('buffer').Buffer ;

data = [ ... Buffer.from('{"method":"a", "body":"testing 111"}\n'), ... Buffer.from('{"method":"B", "body":"testing 111"}\n') ];

n2.parseData(data) ;

//if you want to force send message to listeners:

n2._dispatchToListeners("B", "message here")
n2._dispatchToListeners("B", "message here")
n2._dispatchToListeners("B", "message here")



const n3 = new DispatchMessage2Listeners( 10, (data)=>{
    let parsedString = JSON.parse( data.toString() ) ;
    return { header:parsedString.method, message:parsedString.body }
} ) ;
var objGroup = {__id:23234234}
n3.addListener("B",  (d)=>{
    console.log("recived B again in n3", d) ;
}, objGroup ) ;

n3.addListenerOnChange("B",  (d)=>{
    console.log("recived B once in n3", d) ;
}, objGroup ) ;

n3._dispatchToListeners("B", "another message again") ;
n3._dispatchToListeners("B", "another message again") ;
n3._dispatchToListeners("B", "another message again") ;
n3.removeListenersByGroup( objGroup ) ;
//just 3 first message neet to recive response
n3._dispatchToListeners("B", "another message again") ;
n3._dispatchToListeners("B", "another message again") ;
n3._dispatchToListeners("B", "another message again") ;

```



CHANGELOG.md
## CHANGE LOG

# 1.0.1

Now has a method to listen all

# 1.2.0

Now has a method to listen all and all to merge. Changed logic to add to all, if method name is "*" is equal to add all

# 1.2.1

BUG FIX - Data raw is not sended to listeners, just 

# 2.0.0

CHANGE - Change de responsibility to parse data and parse stream to out of class. For use in another circunstances.


index.js

const pushToMap = (map, key, client) => {
    if( ! map.has( key ) ){
        map.set( key, new Map() ) ;
    }
    const value = map.get(key);
    value.set(client, true) ;
}
const dispatchToList = (list, data)=>{
    if(!list || list.size == 0){
        return ;
    }
    list.forEach((v,k)=>{
        if( v ){
            k(data);
        }
    }) ;
}
const pushToGroup = (listenerList, groupKey, value)=>{
    let itens = [] ;
    if( listenerList.has(groupKey) ){
        itens = listenerList.get(groupKey) ;
    }
    itens.push(value)
    listenerList.set(groupKey, itens) ;
}

class DispatchMessage2Listeners {
    /**
     * 
     * bytes to separate messages
     * @param {*} p_bytesSeparetor 
     * parser method whos get data and returns {header, message}
     * @param {*} parser 
     */
    constructor(){
        //checking parser interface
        this._lastMessages          = new Map() ;
        
        this._listeners             = new Map() ;
        this._listenersMerge        = new Map() ;

        this._listenersGroup        = new Map() ;
        this._logOn                 = false ;
    }
   
    set showLog(v){
        this._logOn = v ;
    }
    //override this if you need change the way to compare
    compare(v1, v2){
        if(v1 === undefined && v2 === undefined){
            return true ;
        }
        if(v1 === undefined  || v2 === undefined ){
            return false ;
        }
        return JSON.stringify(v1) == JSON.stringify(v2) ;
    }
    /**
     * When you get message, call this method.
     * It will send to listener data based in key
     * @param {*} key 
     * @param {*} data 
     */
    dispatchToListeners( key, data ){
        if(!this._listeners.has(key) && !this._listenersMerge.has(key) && !this._listeners.has("*") && !this._listenersMerge.has("*") ){
            if(this._logOn) console.log("has no listener for ", key) ;
            return ;
        }
        var last = this._lastMessages.get(key) ;
        var newstr = data ;
        if( ! this.compare( newstr , last ) ){
            dispatchToList( this._listenersMerge.get("*"), data ) ;
            //is new
            last = newstr;
            this._lastMessages.set(key, last);
            //dispatch to merges
            if(this._logOn) console.log("NEW message")
            dispatchToList( this._listenersMerge.get(key), data ) ;
        } else {
            if(this._logOn) console.log("repeated message")
        }
        dispatchToList( this._listeners.get("*"), data ) ;
        dispatchToList( this._listeners.get(key), data ) ;
    }
    addListenerAll( listener, groupKey = null ){
        return this.addListener("*", listener, groupKey ) ;
    }

    addListenerAllOnChange( listener, groupKey = null ){
        return this.addListenerOnChange("*", listener, groupKey ) ;
    }
    /**
     * Call when get message with method like methodName
     * @param {*} methodName 
     * @param {*} listener 
     */
    addListener( methodName, listener, groupKey = null ){
        if(groupKey){
            //add to a group to posibility remove many listener from group
            pushToGroup(this._listenersGroup, groupKey, {methodName, listener}) ;
            
        }
        pushToMap( this._listeners, methodName, listener ) ;
    }
    /**
     * Call when get message with method like methodName AND value has change
     * @param {*} methodName 
     * @param {*} listener 
     */
    addListenerOnChange( methodName, listener, groupKey = null ){
        if(groupKey){
            //add to a group to posibility remove many listener from group
            pushToGroup(this._listenersGroup, groupKey, {methodName, listener}) ;
            
        }
        pushToMap( this._listenersMerge, methodName, listener ) ;
    }
    /**
     * Remove listener my method name and listener method
     * @param {*} methodName 
     * @param {*} listener 
     */
    removeListener( methodName, listener ){
        if( ! ( this._listenersMerge.has(methodName) ? this._listenersMerge.get(methodName).delete( listener ) : false ) ) {
            return this._listeners.has(methodName)? this._listeners.get(methodName).delete( listener ) : false ;
        }
        return false ;
    }
    /**
     * grouped listeners can be all removed using this method
     * @param {*} group 
     */
    removeListenersByGroup( group ){
        if( this._listenersGroup.has( group ) ){
            let listToRemove = this._listenersGroup.get(group) ;
            listToRemove.forEach((item)=>{
                this.removeListener( item.methodName, item.listener ) ;
            })
        }
    }
}

module.exports = DispatchMessage2Listeners ;

example.js
const DispatchMessage2Listeners = require('./index') ;

const NmeaBuffer = require( 'nmea-buffer' ) ;
const Stream2Event = require( 'socketstream2event' ) ;

//Example 1, parsing NMea messages
const n1 = new DispatchMessage2Listeners() ;

let s2e = new Stream2Event( 10 ) ;
s2e.addOnData((data)=>{
    let nmeaMessage = NmeaBuffer.parse(data) ;
    n1.dispatchToListeners( nmeaMessage.header, nmeaMessage ) ;
})

n1.showLog = true ;
n1.addListener("a", (d)=>{
    console.log("\n\n*1 recived A in n1", d.message) ;
} ) ;

n1.addListener("B", (d)=>{
    console.log("\n\n*2 recived B in n1", d.message) ;
} ) ;
n1.addListenerAll((d)=>{
    console.log("\n\n*3 recived becouse listener all message in n1", d.message) ;
} ) ;
n1.addListenerAllOnChange((d)=>{
    console.log("\n\n*4 recived listen all if CHANGE in n1", d.message) ;
} ) ;
let data = [ ... NmeaBuffer.getMessageBuffer("a", "just for A, test 1"), ...NmeaBuffer.getMessageBuffer("a", "just for A test 2"), ...NmeaBuffer.getMessageBuffer("B", "just for Bs"), ...NmeaBuffer.getMessageBuffer("c", "just for c"), ...NmeaBuffer.getMessageBuffer("c", "just for c") ];

s2e.parseData(data) ;

console.log("\n End test n1. \n ................................. \n\n") ;

//Exemple 2 - parsing using json with method and body attributes
const n2 = new DispatchMessage2Listeners() ;
s2e.addOnData((data)=>{
    let nmeaMessage = NmeaBuffer.parse(data) ;
    n2.dispatchToListeners( nmeaMessage.header, data ) ;
})
n2.addListener("*",  (d)=>{
    console.log("\n\n*recived * in n2", NmeaBuffer.parse(d)) ;
} )

n2.addListenerOnChange("B",  (d)=>{
    console.log("\n\n*recived B in n2", NmeaBuffer.parse(d)) ;
} )
const Buffer        = require('buffer').Buffer ;

data = [ ... Buffer.from('{"method":"a", "body":"testing 111"}\n'), ... Buffer.from('{"method":"B", "body":"testing 111"}\n') ];

s2e.parseData(data) ;

//if you want to force send message to listeners:

n2.dispatchToListeners("B", "message here")
n2.dispatchToListeners("B", "message here")
n2.dispatchToListeners("B", "message here")

console.log("\n End test n2. \n ................................. \n\n") ;



const n3 = new DispatchMessage2Listeners() ;
var objGroup = {__id:23234234}

n3.addListener("B",  (d)=>{
    console.log("recived B again in n3", d) ;
}, objGroup ) ;

n3.addListenerOnChange("B",  (d)=>{
    console.log("recived B once in n3", d) ;
}, objGroup ) ;

n3.dispatchToListeners("B", "another message again") ;
n3.dispatchToListeners("B", "another message again") ;
n3.dispatchToListeners("B", "another message again") ;
n3.removeListenersByGroup( objGroup ) ;
//just 3 first message neet to recive response
n3.dispatchToListeners("B", "another message again") ;
n3.dispatchToListeners("B", "another message again") ;
n3.dispatchToListeners("B", "another message again") ;

console.log("\n End test n3. \n ................................. \n\n") ;

console.time("n4")
const n4 = new DispatchMessage2Listeners() ;
//when you send object to listeners and need change way to compare:
n4.xcompare = (n1,n2)=>{
    if(!n1 && !n2){
        return true ;
    }
    if(! n1 || !n2 ){
        return false ;
    }
    return (n1.m == n2.m)
}
n4.addListener("C",  (d)=>{
    console.log("recived C again in n4", d) ;
}, objGroup ) ;

n4.addListenerOnChange("C",  (d)=>{
    console.log("recived C once in n4", d) ;
}, objGroup ) ;

for( var i = 0; i < 4; i++){
    n4.dispatchToListeners("C", {m:"another message again"}) ;
}


console.timeEnd("n4")
console.log("\n End test n4. \n ................................. \n\n") ;
